app/forms/changeGroupLead/NewGroupLeadEnterContactAddressFormProvider.scala

```diff
@@ -16,13 +16,71 @@
-    Form("value" -> text)
-}
+import javax.inject.Inject
+
+//todo: add more validation: see SoT
+class NewGroupLeadEnterContactAddressFormProvider @Inject() extends Mappings {
+
+   def apply(): Form[NewGroupLeadAddressDetails] = Form(
+     mapping(
+       addressLine1 -> text("newGroupLeadEnterContactAddress.error.addressLine.required")
+        .verifying(maxLength(maxAddressLineLength, "newGroupLeadEnterContactAddress.error.addressLine1.length"))
+         .verifying(isValidAddressLine("newGroupLeadEnterContactAddress.error.addressLine1.invalid.line")),
+       addressLine2 -> text("newGroupLeadEnterContactAddress.error.addressLine.required")
+        .verifying(maxLength(maxAddressLineLength, "newGroupLeadEnterContactAddress.error.addressLine2.length"))
+         .verifying(isValidAddressLine("newGroupLeadEnterContactAddress.error.addressLine2.invalid.line")),
+       addressLine3 -> optional(
+         text("newGroupLeadEnterContactAddress.error.AddressLine2.required")
+           .verifying(maxLength(maxAddressLineLength, "newGroupLeadEnterContactAddress.error.addressLine3.length"))
+           .verifying(isValidAddressLine("newGroupLeadEnterContactAddress.error.addressLine3.invalid.line"))
+       ),
+       addressLine4 -> text("newGroupLeadEnterContactAddress.error.addressLine4.required")
+         .verifying(maxLength(maxAddressLineLength, "newGroupLeadEnterContactAddress.error.addressLine4.length"))
+         .verifying(isValidAddressLine("newGroupLeadEnterContactAddress.error.addressLine4.invalid.line")),
+       postalCode -> optional(
+         text("newGroupLeadEnterContactAddress.error.postalCode.required ")
+           .verifying(maxLength(8, "newGroupLeadEnterContactAddress.error.postalCode.inRange"))
+           .verifying(minLength(5, "newGroupLeadEnterContactAddress.error.postalCode.inRange"))
+           .verifying(isValidAddressLine("newGroupLeadEnterContactAddress.error.postalCode.inRange"))
+       ),
+       countryCode -> text("newGroupLeadEnterContactAddress.error.countryCode.required")
+         .verifying(maxLength(maxAddressLineLength, "newGroupLeadEnterContactAddress.error.countryCode.length"))
+         .verifying(isValidAddressLine("newGroupLeadEnterContactAddress.error.countryCode.invalid"))
+    )(NewGroupLeadAddressDetails.apply)(NewGroupLeadAddressDetails.unapply).verifying(
+       "newGroupLeadEnterContactAddress.error.postalCode.required",
+       fields =>
+         fields match {
+           case address => validatePostalCode(address.countryCode, address.postalCode)
+         }
+     )
+   )
+
+  private def isValidAddressLine(errorKey: String): Constraint[String] =
+    pattern(addressLineRegExp.r, error = errorKey)
+
+  private def validatePostalCode(countryCode: String, postalCode: Option[String]): Boolean = {
+    if(countryCode.equals("GB") && !postalCode.isDefined) false
+    else true
+
+  }
+ }
 
 object NewGroupLeadEnterContactAddressFormProvider {
+  val addressLine1 = "addressLine1"
```

Comment:
can use these vals in the view to tie them together so we know they match

Reply:
yes,  I would need to go back over when refactoring